---
 usb-linux.c |    4 ++++
 1 file changed, 4 insertions(+)

--- qemu-0.12.3.orig/usb-linux.c
+++ qemu-0.12.3/usb-linux.c
@@ -446,6 +446,10 @@ static void usb_host_handle_reset(USBDev
 
     dprintf("husb: reset device %u.%u\n", s->bus_num, s->addr);
 
+    if (((s->descr[8] << 8) | s->descr[9]) == 0x2471 &&
+        ((s->descr[10] << 8) | s->descr[11]) == 0x0853)
+        return;
+
     ioctl(s->fd, USBDEVFS_RESET);
 
     usb_host_claim_interfaces(s, s->configuration);
